### DISCUSSION TITLE
Sharing types and utils for both app and server dirs

### DISCUSSION THREAD

**Discussion by bnachtweh** [nuxt/nuxt#28374] [Ideas]
So I was checking out the Nuxt 4 features and started to experiment. One thing I noticed with the new folder structure (I'm using the `app` folder) is that my imports didn't work as I would expect anymore. The reason I use imports is that I often utilize a global / shared folder of types that are shared between client and server code. I would like this types folder to be shipped with the generated types on `nuxt generate`, so I don't have to import them specifically.

Say for instance I have this nuxt.config.ts:

```
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  compatibilityDate: "2024-08-01",
  future: {
    compatibilityVersion: 4,
  },
  devtools: { enabled: true },
  imports: {
    dirs: ["types"],
  },
});
```

In my opinion that would make sense, as my nuxt config file is in the root of the project. In the "old" setup this also works, because everything but the server code is in the root of the project (I would assume that this is the reason why this currently works). However, with the new `app` folder, the contents of the types folder is not included in the generated imports.

If I change my nuxt config to:

```
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  compatibilityDate: "2024-08-01",
  future: {
    compatibilityVersion: 4,
  },
  devtools: { enabled: true },
  imports: {
    dirs: ["../types"],
  },
});
```

The generated types are visible in `.nuxt/imports.d.ts`, but I would expect to not have to move a folder up in my nuxt config for this. I also would expect that it is possible to include root folders (outside either `app` or `server`) within my imports to make stuff like global / shared types available without having to import them in my application. 

Here's a CodeSandbox to view my project structure: https://codesandbox.io/p/github/bnachtweh/nuxt-sandbox/main

How do you guys feel about this?

**Comment by dencs08** [comment]
I've also stumbled upon this but I believe its default behaviour as stated in the [migration guide](https://nuxt.com/docs/getting-started/upgrade) -> "the new Nuxt default srcDir is app/ by default, and most things are resolved from there.". 
It seems a little inconvenient for me but perhaps there is some deeper thought to it i'm not yet seeing yet.

**Reply by bnachtweh** [reply]
Yeah good point @dencs08. I'm replying to (hopefully) re-address this. Also a thing that could be plausible is that you have shared util functions (next to shared types), that get used both on the server as the client side. It would make sense to me that this util folder could be kept in the root of the project and that both `server` code and `app` code can make use of these shared utils. 

The new approach states that both the `app` folder and the `server` folder can have utils folders, which is fine. As long as I can have an easy configurable place to have shared utils and shared types.

**Comment by andrescellini** [comment]
struggle to figure this out. I was expecting that the import dirs was "./types" since nuxt.config is on the root.

**Reply by bnachtweh** [reply]
Yes, me too! And also because I consider my (root) types folder to be shared in both the client and the server code. 

By the way, I also had this problem when running my Vitest test suite with coverage reporting turned on. The coverage directory would generate inside the app folder, which is unexpected in my opinion.

**Comment by bnachtweh** [comment]
Following up on this; I noticed there is a `shared` folder option now! Exactly what I needed. Closing this discussion...

**Reply by renardsas** [reply]
> Following up on this; I noticed there is a `shared` folder option now! Exactly what I needed. Closing this discussion...

I'm adding the link to the documentation :
[Shared folder](https://nuxt.com/docs/guide/directory-structure/shared)
